### ISSUE TITLE
Please put the ::Error class in a real namespace or rename it something like "::HeifError"

### ISSUE THREAD

**Issue by danhd123** [strukturag/libheif#1125]
"Error" is a very big claim in the "::" namespace, and it conflicts at link time with other libraries, some of which (eg Helix's Perforce API) are effectively un-recompilable. While it's not fair that libheif should be the one to fix this particular conflict, no library should be polluting the :: namespace like this anyway. 

I have had to remove heif support from my static ImageMagick build (which likely means even more work later) because of this.

**Comment by bradh** [comment]
Do you mean in the C++ bindings?

If so, it looks like it already is:
https://github.com/strukturag/libheif/blob/33e00a4ec54e6fffca3febe3054017b1b81a0c49/libheif/heif_cxx.h#L34-L36

Can you show a small example of the problem?

**Comment by danhd123** [comment]
https://github.com/strukturag/libheif/blob/33e00a4ec54e6fffca3febe3054017b1b81a0c49/libheif/error.h#L70

This `Error` class is not in the heif namespace.

**Comment by bradh** [comment]
> https://github.com/strukturag/libheif/blob/33e00a4ec54e6fffca3febe3054017b1b81a0c49/libheif/error.h#L70
> 
> This `Error` class is not in the heif namespace.

That isn't part of the public API AFAICT - it isn't an installed header.

**Comment by danhd123** [comment]
That's possible. I'll see if I can find the flags ImageMagick uses.

**Comment by danhd123** [comment]
Here are the flags that Visual Studio is using to compile libheif for ImageMagick:
```/JMC /MP /ifcOutput "Debug\StaticMTD-x64\CORE_heif\" /GS /W0 /Gy /Zc:wchar_t /I"..\..\..\aom" /I"..\..\..\de265" /I"..\..\..\heif" /Zi /Gm- /Od /Ob0 /Fd"..\..\bin\CORE_DB_heif_.pdb" /Zc:inline /fp:precise /D "_DEBUG" /D "_WINDOWS" /D "WIN32" /D "_VISUALC_" /D "NeedFunctionPrototypes" /D "ENABLE_PARALLEL_TILE_DECODING" /D "HAVE_AOM_DECODER" /D "HAVE_AOM_ENCODER" /D "HAVE_LIBDE265" /D "LIBDE265_STATIC_BUILD" /D "LIBHEIF_STATIC_BUILD" /D "_LIB" /D "_MBCS" /errorReport:prompt /GF /WX- /Zc:forScope /RTC1 /std:c17 /Gd /Oy- /MDd /openmp- /std:c++17 /FC /Fa"Debug\StaticMTD-x64\CORE_heif\" /EHsc /nologo /Fo"Debug\StaticMTD-x64\CORE_heif\" /Fp"Debug\StaticMTD-x64\CORE_heif\CORE_DB_heif_.pch" /diagnostics:column```

**Comment by bradh** [comment]
Sorry, I meant what is going into the cmake step, rather than what is coming out.

There is a CMake flag "WITH_REDUCED_VISIBILITY" which should default to ON.  

That said, I'm not sure what it does in Microsoft's compiler. With clang and gcc, it does `-fvisibility=hidden`, and internal symbols are not accessible. Possibly it is interacting with the debug options, but I'm hand-waving at this point.

**Comment by danhd123** [comment]
There is no CMake configuration, it's being configured by VisualMagick. The config file for that is as follows:

```[DLL]

[LICENSE]
COPYING

[DEFINES]
ENABLE_PARALLEL_TILE_DECODING
HAVE_AOM_DECODER
HAVE_AOM_ENCODER
HAVE_LIBDE265
LIBDE265_STATIC_BUILD

[DEFINES_LIB]
LIBHEIF_STATIC_BUILD

[DEFINES_DLL]
LIBHEIF_EXPORTS

[DIRECTORIES]
heif\libheif
heif\libheif\color-conversion
heif\libheif\plugins

[INCLUDES]
aom
de265
heif

[EXCLUDES]
decoder_dav1d.cc
decoder_ffmpeg.cc
decoder_jpeg.cc
decoder_openjpeg.cc
encoder_rav1e.cc
encoder_jpeg.cc
encoder_openjpeg.cc
encoder_kvazaar.cc
encoder_svt.cc
encoder_x265.cc
plugins_*

[DEPENDENCIES]
aom
de265

[CONFIG_DEFINE]
/*
  Define to use the libheif library
*/
#define MAGICKCORE_HEIC_DELEGATE
```

**Comment by bradh** [comment]
Can you check what happens if you use cmake without the ? I'm trying to work out if this is a libheif issue or a problem in that build system you are using.

**Comment by silverbacknet** [comment]
IIRC by modern C++ rules (since C++11), this should all be solved without any flags needed by enclosing the header in an anonymous namespace:

```
namespace {
class Error {
....
}
```

The users of the class shouldn't have to change at all, anonymous is used the same way as no namespace, except that it's either internal linkage only or name-mangled to prevent collisions, depending on the compiler.

**Comment by danhd123** [comment]
I generated the project with `cmake -DBUILD_SHARED_LIBS=OFF .`, and built libheif in Visual Studio 2022 (2019 toolchain). `dumpbin /symbols` includes the following:

```
32D 00000000 SECT9E notype ()    External     | ??1Error@@QEAA@XZ (public: __cdecl Error::~Error(void))
32E 00000000 SECT77 notype ()    External     | ??0Error@@QEAA@AEBV0@@Z (public: __cdecl Error::Error(class Error const &))
...
B2B 00000000 UNDEF  notype ()    External     | ??0Error@@QEAA@XZ (public: __cdecl Error::Error(void))
```
So I don't think that WITH_REDUCED_VISIBILITY default is working as expected.

**Comment by silverbacknet** [comment]
Yes, there's no switch for it in msbuild. I just double-checked the full command listing to be sure, but there's nothing like -fvisibility=hidden, so only doing it via the code would work.

**Comment by danhd123** [comment]
Looking around StackOverflow, this seems to be a common question around static library builds -- and it looks like there isn't a way to keep internal symbols from being visible (to change the default symbol-hiding behavior in MSVC systems, it's `__declspec(dllexport)` and not `__declspec(libexport)` for a reason). So a possible alternative would be to strongly disrecommend or disallow building libheif statically, but I think that's overkill when wrapping the internal classes in a namespace would fix the issue.
